[Messenger] Introduce `DeduplicateMiddleware`

VincentLanglet opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Issues        | Fix #54126  <!-- prefix each issue number with "Fix #", no need to create an issue if none exists, explain below instead -->
| License       | MIT

jderusse commented on src/Symfony/Component/Messenger/Message/LockableMessage.php:
```diff
@@ -0,0 +1,22 @@
+    /**
+     * Returns null if you want to force the dispatch of the message.
+     */
+    public function getKey(): ?Key;
```
I think this should be `function getLockKey(): ?string;`

The complexity (and coupling with symfony/lock) of generating the Key, could be handled by the Middleware.
That would provide a simpler/easier to implement interface for end-users.

jderusse commented on src/Symfony/Component/Messenger/Middleware/LockMiddleware.php:
```diff
@@ -0,0 +1,68 @@
+            if (null !== $key) {
+                // The acquire call must be done before stamping the message
+                // in order to have the full state of the key in the stamp.
+                $canAcquire = $this->lockFactory->createLockFromKey($key, autoRelease: false)->acquire();
```
IMHO LockTTL should be configurable

VincentLanglet replied:
Should it be another interface to implement ?
-> If you implements LockMessage => you have the default TTL
-> If you implements something like ConfigurableLockMessage => You have to return a custom TTL 

Or should both method be in the LockMessage interface ?

jderusse replied:
I don't have strong opinion on this (the same question for the other comment about `release before handle` thing)

VincentLanglet replied:
I think:
- 3 interfaces would be too much
- TTL might not be often configured

So I tried with two interfaces:
- LockableMessageInterface
- TTLAwareLockableMessageInterface

jderusse commented on src/Symfony/Component/Messenger/Message/LockableMessage.php:
```diff
@@ -0,0 +1,22 @@
+
+use Symfony\Component\Lock\Key;
+
+interface LockableMessage
```
```suggestion
interface LockableMessageInterface
```

jderusse commented on src/Symfony/Component/Messenger/Middleware/LockMiddleware.php:
```diff
@@ -0,0 +1,68 @@
+            if (null !== $envelope->last(ReceivedStamp::class)) {
+                $stamp = $envelope->last(LockStamp::class);
+                if ($stamp instanceof LockStamp) {
+                    $this->lockFactory->createLockFromKey($stamp->getKey(), autoRelease: false)->release();
```
I wonder if we should provide an option to release the lock BEFORE calling the handler.

Use case is: My Handler takes time to read data THEN process them. If data changed, I consider that a new message should be dispatched and queued even if the previous message is not totally handled.

VincentLanglet replied:
That's a good idea. (Just need to keep in my that it would mean that again we cannot know if the message will be sucessfull or retried. Cf https://github.com/symfony/symfony/pull/54141#discussion_r1522693445)

Should it be the same interface
```
interface LockableMessageInterface
{
    public function getKey();

    public function releaseBefore();
}
```
or multiple interfaces ?

jderusse commented on src/Symfony/Component/Messenger/Middleware/LockMiddleware.php:
```diff
@@ -0,0 +1,68 @@
+
+        try {
+            $envelope = $stack->next()->handle($envelope, $stack);
+        } finally {
```
What if message failed and is retried? :thinking:

VincentLanglet replied:
Failure/Retry are handled by the SendFailedMessageForRetryListener
It listen this event https://github.com/symfony/symfony/blob/3fdc2c99d101ad69dd5c99ed03d8d64ad9608631/src/Symfony/Component/Messenger/Worker.php#L199 

But I "handle" the lock before
https://github.com/symfony/symfony/blob/3fdc2c99d101ad69dd5c99ed03d8d64ad9608631/src/Symfony/Component/Messenger/Worker.php#L163

So I don't know at the moment if the message will be retried or not...

I prefer to release the lock rather than riskying keeping it forever because the message is retry. I don't see other solution so far. There was already another chance to have some duplication in the queue, if the first lock reach the end of the TTL. We cannot be sure there is not a big waiting line in the queue some time...

I don't see the LockMiddleware as a way to have absolute 0 duplication in the queue but as a way to reduce the duplication as much as technically possible. But I'm open to any suggestion to improve this.

jderusse replied:
I agree.

jderusse commented on src/Symfony/Component/Messenger/Middleware/LockMiddleware.php:
```diff
@@ -0,0 +1,68 @@
+            if (null !== $key) {
+                // The acquire call must be done before stamping the message
+                // in order to have the full state of the key in the stamp.
+                $canAcquire = $this->lockFactory->createLockFromKey($key, autoRelease: false)->acquire();
```
Is there a mechanism that asserts the key can be serialized?
If people created a lockFactory with stores like flock, I'm pretty sure it would not work as expected (The key contains a file resource and won't be serialized)

VincentLanglet replied:
Indeed, seems like store like FlockStore is an issue since they call `Key::markUnserializable`.
But like some other store, if we don't have the state of the key we cannot release it
https://github.com/symfony/symfony/blob/3fdc2c99d101ad69dd5c99ed03d8d64ad9608631/src/Symfony/Component/Lock/Store/FlockStore.php#L131-L133

Do you see a solution for this ? Or should this middleware be limited to serializable key ?

jderusse replied:
maybe their is nothing to do, and just let people deal with the error `UnserializableKeyException`.

If we want to be smarter, we would need to add an interface marker on Store and provide a method on the factory (or lock generated by the factory), but I'm not sure it worth it

jderusse reviewed (commented):
I like the idea, and using the lock component provides a simple solution.

I wonder if this should be named `LockMiddleware` or `DedupplicationMiddleware`. At first I though this PR was about avoid processing similar messages in parallel.

jderusse commented on src/Symfony/Component/Messenger/Middleware/LockMiddleware.php:
```diff
@@ -0,0 +1,88 @@
+                    // The acquire call must be done before stamping the message
+                    // in order to have the full state of the key in the stamp.
+                    $lock = $message instanceof TTLAwareLockableMessageInterface
+                        ? $this->lockFactory->createLock($key, $message->getTTL(), autoRelease: false)
```
```suggestion
                        ? $this->lockFactory->createLockFromKey($key, $message->getTTL(), autoRelease: false)
```

jderusse commented on src/Symfony/Component/Messenger/Middleware/LockMiddleware.php:
```diff
@@ -0,0 +1,88 @@
+                    // in order to have the full state of the key in the stamp.
+                    $lock = $message instanceof TTLAwareLockableMessageInterface
+                        ? $this->lockFactory->createLock($key, $message->getTTL(), autoRelease: false)
+                        : $this->lockFactory->createLock($key, autoRelease: false);
```
same

jderusse commented on src/Symfony/Component/Messenger/Middleware/LockMiddleware.php:
```diff
@@ -0,0 +1,88 @@
+                        : $this->lockFactory->createLock($key, autoRelease: false);
+                    $canAcquire = $lock->acquire();
+
+                    $envelope = $envelope->with(new LockStamp($key, $message->shouldBeReleasedBeforeHandlerCall()));
```
shouldn't it be moved after the check on `if (!$canAquire)`?

What's the purpose of adding the stamp of we don't process it?

VincentLanglet replied:
Yes indeed

jderusse commented on src/Symfony/Component/Messenger/Middleware/LockMiddleware.php:
```diff
@@ -0,0 +1,90 @@
+        $message = $envelope->getMessage();
+
+        if (null === $envelope->last(ReceivedStamp::class)) {
+            if ($message instanceof LockableMessageInterface) {
```
what's about moving this line with the first `if` at line 34?
It would avoid processing all the logic if the message should not be locked.

VincentLanglet replied:
Done

jderusse commented on src/Symfony/Component/Messenger/Middleware/LockMiddleware.php:
```diff
@@ -0,0 +1,90 @@
+            $envelope = $stack->next()->handle($envelope, $stack);
+        } finally {
+            // If we've received a lockable message, we're releasing it.
+            if (null !== $envelope->last(ReceivedStamp::class)) {
```
I think this is not necessary, the check is performed in the releaseLock method.
Also, it's not consistent with the call `$this->releaseLock($envelope, true);` at line 64

VincentLanglet replied:
For `$this->releaseLock($envelope, true);` the check on the `ReceivedStamp` is done because I do:
```
if (null === $envelope->last(ReceivedStamp::class)) {
    // foo
} else {
    $this->releaseLock($envelope, true);
}
```

And `releaseLock` is not doing the any check on `ReceivedStamp`, but on `LockStamp`.
If I don't do the `ReceivedStamp` check, I will release the lock right after adding the lock (during the first dispatch) instead of when I receive the message.

jderusse commented on src/Symfony/Component/Messenger/Middleware/LockMiddleware.php:
```diff
@@ -0,0 +1,90 @@
+        $stamp = $envelope->last(LockStamp::class);
+        if ($stamp instanceof LockStamp && $stamp->shouldBeReleasedBeforHandlerCall() === $beforeHandlerCall) {
+            $message = $envelope->getMessage();
+            $lock = $message instanceof TTLAwareLockableMessageInterface
+                ? $this->lockFactory->createLockFromKey($stamp->getKey(), $message->getTTL(), autoRelease: false)
+                : $this->lockFactory->createLockFromKey($stamp->getKey(), autoRelease: false);
+            $lock->release();
```
we want to release the lock, no need to set `autoRelease: false` nor setting the TTL.

```suggestion
            $lock = $this->lockFactory->createLockFromKey($stamp->getKey())
            $lock->release();
```

jderusse commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -2130,6 +2135,10 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
                 ['id' => 'handle_message'],
             ],
         ];
+        if (class_exists(LockMiddleware::class)) {
```
we could also check for the existence of class LockFactory :thinking: 
If the lock component is not installed, let's avoid this by adding an extra middleware.

And LockFactory could be required in constructor's arguments

VincentLanglet replied:
Indeed, done

ro0NL commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -2130,6 +2135,10 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
             ],
         ];
+        if (class_exists(LockMiddleware::class)) {
+            $defaultMiddleware['before'][] = ['id' => 'lock_middleware'];
```
im not sure about this default, adding it in userland is little effort

edit: oh sry, the middleware does nothing by default. It's fine i guess.

ro0NL commented on src/Symfony/Component/Messenger/Middleware/LockMiddleware.php:
```diff
@@ -0,0 +1,82 @@
+    public function handle(Envelope $envelope, StackInterface $stack): Envelope
+    {
+        $message = $envelope->getMessage();
+        if (!$message instanceof LockableMessageInterface) {
```
i'd prefer putting the key in a stamp

VincentLanglet replied:
WDYT @jderusse ?

it's
```
class MyMessage implements LockableMessageInterface
{
    public function getKey() { ... } 
}
```
vs
```
$this->dispatch(new MyMessage(), new LockStamp(new Key('myKey')));
```

I thought it was better in the Message, because the key will be often related to the data in the message.
For instance
```
class MyMessage implements LockableMessageInterface
{
    public construct __construct(public int projectId) {}
    public function getKey() { return self::class . '_' . $this->projectId } 
}
```

ro0NL replied:
the stamp brings infinite runtime flexibility without any coupling

adding them by default to messages can be middleware/service concern :) or wait for SF to allow default stamping thru stamps-as-attributes

jderusse replied:
I'm :+1: with `LockableMessageInterface`.

From userland point of view it's simpler to only return a string

It avoids coupling with symfony/lock (people don't have to know how works the middleware)

Also, if the message is dispatched from multiple places, you have to remember that the envelope should contain a stamp.

VincentLanglet replied:
I tried the implementation with the Stamp only with
https://github.com/symfony/symfony/pull/54141/commits/c2331215c9c89d499c7a7b42b9eed44afb873051

And it has some advantages like less code needed (and 0 interfaces) and maybe more flexibility.

We don't provide a `DelayedMessageInterface` to delay message but a DelayStamp instead.
It could be consistent to not provider an interface but just a stamp.

Maybe another feature should be to provide a `StampedMessageInterface` with a method `getDefaultStamps()`.
Other people can have this behavior by doing
```
class MyMessage
{
    public function toEnvelope(): Envelope
    {
        return new Envelope($this, $defaultStamps);
    }
}
```
and writing `$bus->dispatch((new MyMessage())->toEnvelope())`.

Anyway, it's in a separate commit so I can revert/keep it, like you want.

VincentLanglet commented:
Friendly ping @jderusse @ro0NL do you have time for new look ?

welcoMattic commented on src/Symfony/Component/Messenger/Middleware/LockMiddleware.php:
```diff
@@ -0,0 +1,57 @@
+
+final class LockMiddleware implements MiddlewareInterface
+{
+    private LockFactory $lockFactory;
+
+    public function __construct(LockFactory $lockFactory)
+    {
+        $this->lockFactory = $lockFactory;
+    }
```
We can use CPP here

VincentLanglet replied:
Done

ro0NL commented on src/Symfony/Component/Messenger/Stamp/LockStamp.php:
```diff
@@ -0,0 +1,46 @@
+
+    public function __construct(
+        string $key,
+        ?float $ttl = 300.0,
```
this couples with an unknown handling time

i think it requires #53508

VincentLanglet replied:
> this couples with an unknown handling time
> 
> i think it requires #53508

#53508 might provide an improvement, but is not a blocker.
As a proof, it's been month I'm using this LockStamp without issues.

- If your handling time is short < 300s, there is no issue
- If your handling time is longer, you can still provide a bigger ttl
- If your handling time is longer, you can also consider that if it's been 300s the process started, and it's okay to add a new one in the queue. (If it's the issue you think it can happen ; I'm not even sure I encountered it so far)

At least the LockStamp provide a way to handle race-condition and avoid duplicate in a duration-given (the ttl).

We could consider a merge as-if and another PR after #53508 to improve the feature in a way you have in mind (?)

ro0NL commented on src/Symfony/Component/Messenger/Middleware/LockMiddleware.php:
```diff
@@ -0,0 +1,54 @@
+                return $envelope;
+            }
+        } elseif ($stamp->shouldBeReleasedBeforeHandlerCall()) {
+            $this->lockFactory->createLockFromKey($stamp->getKey())->release();
```
im not sure i understand this

ro0NL commented on src/Symfony/Component/Messenger/Middleware/LockMiddleware.php:
```diff
@@ -0,0 +1,54 @@
+        }
+
+        if (null === $envelope->last(ReceivedStamp::class)) {
+            $lock = $this->lockFactory->createLockFromKey($stamp->getKey(), $stamp->getTtl(), autoRelease: false);
```
are we sure about `autoRelease: false` as default

ro0NL commented on src/Symfony/Component/Messenger/Stamp/LockStamp.php:
```diff
@@ -0,0 +1,46 @@
+    private bool $shouldBeReleasedBeforeHandlerCall;
+
+    public function __construct(
+        string $key,
```
i tend to believe the key should be derived from type+payload

VincentLanglet replied:
I think it can be an improvement later to allow the key to be nullable and autogenerated.

But key should be possible to be passed because you might want/need to use the same key for different payload. For instance let's say you pass the timestamp in the payload but you want to process one message per day, you cannot have the timestamp in the key.

ro0NL replied:
fair

but why put a timestamp in the payload, if it's going to be locked by single key anyway ...

VincentLanglet replied:
> but why put a timestamp in the payload, if it's going to be locked by single key anyway ...

You may want to run some computation one time a day, but log the moment of the day the computation was asked for the first time ; dunno. 
Also, some people may want two differents message to share the same key ?

I have some simple usage of this feature (and use it on my project atm, working fine) but I cannot know every potential usage of this stamp, and prefer to start with simplicity and flexibility.

ro0NL replied:
simplicity = autogenerating the key by default ;)

simplicity is also not having to deal with ttl ;P

VincentLanglet replied:
Simplicity is about starting with the symfony feature which requiert the less code in symfony folder.

And the current implementation is the simplest one which also provide flexibility
- by passing the key you let chose the user the rule about which payload should be locked and which shouldn't.
- ttl is filled by default like every other constructor of the lock component, but you can override it.

Rather than using passive aggressive tone hidden behind some emoji, I'm totally open for your PR changing the key to a nullable one which is autogenerated when none is passed. Nothing in this PR seems to block another improvement later.
Currently, I'd prefer to move step by step on this LockMiddleware. The missing feature was reported more than one year ago and nobody made a move on it https://github.com/symfony/symfony/discussions/49215. And as shown by some reaction on the PR, it will already solve multiple developers issues.

> simplicity = autogenerating the key by default ;)
> simplicity is also not having to deal with ttl ;P

I'm sorry but this is anything but constructive comments in my eyes.

ro0NL replied:
im considering simplicity from the consumer/enduser POV

ideally these issue are dealt with now

VincentLanglet commented:
Hi, maybe would you have time to give me a feedback on this @nicolas-grekas ?

echantigny commented:
I was looking for a way to avoid having duplicate messages in my queue and stumbled on this.  I really wish this feature can be implemented soon.

VincentLanglet commented:
> I was looking for a way to avoid having duplicate messages in my queue and stumbled on this. I really wish this feature can be implemented soon.

I personnally use this implementation on my project and it works fine. You can duplicate it on your projet if needed until the merge.

echantigny commented:
> > I was looking for a way to avoid having duplicate messages in my queue and stumbled on this. I really wish this feature can be implemented soon.
> 
> I personnally use this implementation on my project and it works fine. You can duplicate it on your projet if needed until the merge.

I might give that a try if nothing moves here.  Thanks

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1986,6 +1987,10 @@ private function registerLockConfiguration(array $config, ContainerBuilder $cont
     {
         $loader->load('lock.php');
 
+        if (!class_exists(LockMiddleware::class) || !class_exists(LockFactory::class)) {
```
Doesn't this block need to be moved to the `registerMessengerConfiguration()` method as the definition won't exist if the Messenger component is not configured?

VincentLanglet replied:
Indeed, I moved it.

xabbuh commented on src/Symfony/Component/Messenger/Stamp/LockStamp.php:
```diff
@@ -0,0 +1,46 @@
+
+    public function __construct(
+        string $key,
+        ?float $ttl = 300.0,
```
Let's use constructor property promotion here and for `$shouldBeReleasedBeforeHandlerCall`.

VincentLanglet replied:
Done

xabbuh commented on src/Symfony/Component/Messenger/composer.json:
```diff
@@ -28,6 +28,7 @@
         "symfony/http-kernel": "^6.4|^7.0",
         "symfony/process": "^6.4|^7.0",
         "symfony/property-access": "^6.4|^7.0",
+        "symfony/lock": "^6.4|^7.0",
```
Please add a `"symfony/lock": "<6.4"` entry to the `conflict` section.

VincentLanglet replied:
Done

yariksheptykin replied:
@VincentLanglet @xabbuh  I would like to make use of your `DeduplicateMiddleware` in a legacy symfony v5.4 application. I am wondering why you decided to conflict this package against symfony/lock < v6.4 . Is there anything special in Lock v6.4 that the de-duplication relies on?

I did a quick look at the Lock Component in v5.x. The API of the LockFactory did not change between 5.x and 6.x. To me this Middleware seems compatible with 5.x too.

chalasr replied:
It probably is. We require maintained versions for newly added dependencies.

yariksheptykin replied:
Perfect. Thanks for your quick reply, @chalasr .

ro0NL reviewed (commented):
IIUC the lock ttl should at least include the potential pending queue time

maidmaid commented:
I personally use a very similar implementation, except for the lock mechanism : when you deploy a new application version, you may want to flush the messages queue. In that case, you don't have any out-of-the-box solution to release also the associated locks, so new dispatched messages can be aborted by mistake because the locks are still here.

So, I personally moved from locks to cache (if the key cache is hit, the message isn't dispatched). Thus, I can use `cache:pool:clear` command to also flush the message locks while a deploy.

VincentLanglet commented:
> I personally use a very similar implementation, except for the lock mechanism. When you deploy a new application version, you may want to flush the messages queue. In that case, you don't have any out-of-the-box solution to release also the associated locks, so new dispatched messages can be aborted by mistake because the locks are still here.
> 
> So, I personally moved from locks to cache (if the key cache is hit, the message isn't dispatched). Thus, I can use `cache:pool:clear` command to also flush the message locks while a deploy.

Lock can be implemented different way, and I think there is lock relying on cache ; so it could still be used with lock and you reset your redis or similar cache used.

VincentLanglet commented:
Hi @OskarStark (@jderusse @welcoMattic @xabbuh), any help to move this forward ?

I personally use this implementation and it works fine.
I also got an approval.

maidmaid commented:
> I think there is lock relying on cache

That's right but you'll have to handle flushing on the redis side. Perhaps, `messenger:prune` could be introduced in a next PR to solve this. It would handle flushing messages queue and releasing locks by iterating on each message for looking after a lock stamp.

ro0NL commented:
>  Perhaps, messenger:prune could be introduced in a next PR 

no, single PR per topic pls.

im not sure i understand the cache vs. lock issue at this point.

pounard commented on src/Symfony/Component/Messenger/Middleware/LockMiddleware.php:
```diff
@@ -0,0 +1,54 @@
+    public function handle(Envelope $envelope, StackInterface $stack): Envelope
+    {
+        $stamp = $envelope->last(LockStamp::class);
+        if (!$stamp instanceof LockStamp) {
```
Why instanceof here and not simply `if (!$stamp)` ?

VincentLanglet replied:
More explicit, so clearer.
And it gives better autocompletion.

pounard replied:
And it is probably slower, for autocompletion, you can use an extra assert statement instead which will never be executed on prod!

pounard commented on src/Symfony/Component/Messenger/Middleware/LockMiddleware.php:
```diff
@@ -0,0 +1,54 @@
+
+        if (null === $envelope->last(ReceivedStamp::class)) {
+            $lock = $this->lockFactory->createLockFromKey($stamp->getKey(), $stamp->getTtl(), autoRelease: false);
+            if (!$lock->acquire()) {
```
Why no blocking here? Will it simply be marked for retry? If so, maybe it worth a comment here, since the execution flow is split into many middlewares, it might be good explaining why you do this.

VincentLanglet replied:
A blocking acquire have a lot of drawback since it's slower.
If you're generating a lot of messenger message, your middleware won't follow.
I already got CPU issues in production because of blocking acquire.

But we could introduce a `shouldBeBlocking` parameter to the Stamp if needed.

pounard replied:
Yes that's what I was thinking but it doesn't answer the second question !

VincentLanglet replied:
Rereading the implementation and the param `$blocking`, I don't think there is anything to change/add.

The purpose of the LockMiddleware is to avoid having duplicated similar message in the Queue. There is no reason to wait for the first message to be treated before adding the second message ; that already the purpose of a queue! So the only interesting behavior here is to drop the second message when the first is not treated yet.

pounard replied:
Alright! - I was thinking about use cases where you'd tag messages that for example would create lock contention on the database using the same key: you may want to ensuring that all would pass one after another but all messages must be consumed.*

I happened to implement such locking mechanism using a similar middleware on client projects because of this. In this use case, all messages are valid, you just don't want them pass simultaneously to avoid too many transactions failures and retries.

VincentLanglet replied:
> you just don't want them pass simultaneously

Isnt the principle of a queue to consume them one by one from the queue ?

lermontex replied:
@pounard, I had a similar issue and was hoping that LockMiddleware could solve it.

@VincentLanglet, In case all messages have to be processed strictly in order, simply dropping the message for re-processing will create a mess in the queue even if we manually (in the handler) check if the message should be processed and resend it if necessary.

It would also be nice to be able to customize the strategy of handling repeated messages (resend/reset)

lermontex replied:
@VincentLanglet,

> Isnt the principle of a queue to consume them one by one from the queue ?

A simple example: you are using RabbitMQ and you have multiple handlers - whether you like it or not, messages will be sent for processing one by one, but will be processed in parallel (because you have multiple handlers)

Of course, you can use only one handler, but in that case you will have to create a separate queue for such messages, or use some specific transport settings, but it is not always reasonable.

VincentLanglet replied:
A lock middleware is definetly not the solution for this need. It would block a whole process while waiting for dispatch the new message.

A blocking lock is not a good idea

pounard replied:
> Isnt the principle of a queue to consume them one by one from the queue ?

A bus is not always simply a queue. We always spawn 2 to 4 parallel workers per default when we run a project based upon messenger, or any other message bus.

pounard replied:
> A lock middleware is definetly not the solution for this need. It would block a whole process while waiting for dispatch the new message.

In our own custom implementation, we didn't always block: in first iterations at least, then we fallbacked at some point in a different algo where we resent the message in queue for later consumption. In this scenario, messages were pushed back as if they were just sent, letting other messages that arrived in the meantime pass before them. It worked fine in most cases.

Hence my question about what happens in the lock fails? Is the message sent back for retry?

VincentLanglet replied:
> In our own custom implementation, we didn't always block: in first iterations at least, then we fallbacked at some point in a different algo where we resent the message in queue for later consumption. In this scenario, messages were pushed back as if they were just sent, letting other messages that arrived in the meantime pass before them. It worked fine in most cases.

I'm sorry but I don't understand this dicussion then, since you talk about another feature and another implementation. We both agree the LockMiddleware won't work for your custom implementation, but that's definetly not the purpose.

This middleware is about deduplication and dropping the duplicated message. So if the lock cannot be aquire, the message is dropped. There is no retry queue.

pounard replied:
Yes, but the way you implemented it using a stamp with a key is exactly the way I'd implement my own need.

I guess it would worth it expanding the feature to:

1. allow it to be blocking using a boolean optional parameter on the stamp (won't use it myself, but why not)
2. allow messages to be requeued instead of being dropped using another boolean optional parameter.

With those two minor additions to the stamp, this single middleware should cover all use cases.

Your PR is a very good idea, and I wouldn't want to have a second middleware whereas this one exists and does exactly what I need except for a single line of code.

VincentLanglet replied:
> I guess it would worth it expanding the feature to:

This PR is already opened for 6 month without any changes, and still waiting review.

Let's work with small step.
First merging this.
Then, in another PR adding the new feature you need.

Better is the enemy of good.

But feel free to already open the PR with additional features.

pounard replied:
> Better is the enemy of good.

Agreed. I was thinking that maybe, with a wider range of valid use cases, it might become more interesting to reviewers.

But as you wish. If your PR passes as-is, I'll open new ones when I will experience trouble with locking :smile: 

Thank you for answering.

nicolas-grekas replied:
Wait, this is major concern, I don't think it's resolved at all.
The minimum would be to make the behavior crystal clear: if you add the stamp, messages will be dropped.
Given the discussion, I wonder if the name of the middleware isn't the confusing part.
A lock waits or delays, that's the common expectation.

VincentLanglet replied:
> A lock waits or delays, that's the common expectation.

Why would it be the common expactation ?

In the symfony documentation https://symfony.com/doc/current/components/lock.html there is a lot of example with  the following code
```
$lock = $factory->createLock('pdf-creation');
if (!$lock->acquire()) {
    return;
}
```
and the blocking parameter is false by default.

I would say that a lock waits or skip the task, with a defaut behavior of skipping the task.

nicolas-grekas replied:
The very fact that we have this discussion means there's a topic to clarify. Please mind the rest of my comment.

VincentLanglet replied:
> The very fact that we have this discussion means there's a topic to clarify. Please mind the rest of my comment.

Then, what do you prefer
- A blocking parameter on the lockstamp ?
- Renaming the lockstamp and the lockmiddleware (to what ?)

To me this is very related to the lock behavior since it require the lock component and uses lock key.

nicolas-grekas replied:
Adding a required argument so that people are required to be explicit?
Tentatively:
`$mode = self::MODE_BLOCK|self::MODE_DISCARD`

VincentLanglet replied:
Like this https://github.com/symfony/symfony/pull/54141/commits/72aae9446f52176f74c55f3a405614ba010b864d ?

nicolas-grekas replied:
LGTM, with an `@param` :)

WDYT @pounard?

lermontex replied:
@nicolas-grekas, @VincentLanglet

Note that this implementation blocks the main thread in which the message was sent, so you have to wait, even if you are using asynchronous transport, before you can send the message again. It seems to be correct to block the thread (if necessary) after the message is received (ReceivedStamp) and then take some action

VincentLanglet replied:
That's exactly my point:
- Blocking this lock is not useful in the main thread
- I can add another lock at the reception of the message like [41c0b2f](https://github.com/symfony/symfony/pull/54141/commits/41c0b2f10aa92ee03bcd2a82ddb6e7a324409f98) but we're starting to add lot of feature for different use cases and people might keep continuing asking new mode with usecase delaying this PR. It would be easier to add modes/beahvior one by one. I still tried to had every possible case.

Anyway would something like [41c0b2f](https://github.com/symfony/symfony/pull/54141/commits/41c0b2f10aa92ee03bcd2a82ddb6e7a324409f98)satisfy you @lermontex ?

lermontex replied:
@VincentLanglet, That's better, at least now it's clear what will happen to the message. Thank you!

pounard replied:
> Adding a required argument so that people are required to be explicit?

I'm totally fine with this, we may add a non-blocking-and-requeue behavior later if necessary.

Yet in the use case I had in the past, the non-blocking-and-requeue behavior is the one I needed the most, but if we doesn't reach a consensus about it, then I'm OK to drop it and attempt a new PR in the future.

pounard replied:
> Anyway would something like [41c0b2f](https://github.com/symfony/symfony/pull/54141/commits/41c0b2f10aa92ee03bcd2a82ddb6e7a324409f98)satisfy you @lermontex ?

It seems unnecessary complexity to me, I'd wish to have simply three mode:
- Lock and block.
- Lock and discard on lock failure.
- Lock and re-queue message on lock failure.

All at consumption, not upon dispatch, a dispatched message doesn't guarantee parallel consumption.

With those three modes, I think it should cover all use case:
- Two similar messages received at the same time drops the duplicated ones (lock or discard).
- Two messages than cannot be treated at the same time can block and wait (lock and block).
- Two messages than cannot be treated at the same time under pressure and you don't want a bus thread to be blocked can continue its operation while the message is re-queued (lock and re-queue).

Third scenario seems the most complicated one but the most effective performance wise for the scenario where you have messages that cannot be treated concurrently: treat the lock failure as an error and let the retry algorithm handle it.

The first scenario doesn't completely fix what you meant to fix in the first place though: in the case were you have duplicated messages, it would prevent them from being processed all at once, but it will not prevent them from being processed one after another if somewhat, they're being sent at the same time, but treated at different times because of some other messages dispatched at the very same moment but delaying the processing between the two.

VincentLanglet replied:
> > Anyway would something like [41c0b2f](https://github.com/symfony/symfony/pull/54141/commits/41c0b2f10aa92ee03bcd2a82ddb6e7a324409f98)satisfy you @lermontex ?
> 
> It seems unnecessary complexity to me, I'd wish to have simply three mode:
> 
> * Lock and block.
> * Lock and discard on lock failure.
> * Lock and re-queue message on lock failure.
> 
> All at consumption, not upon dispatch, a dispatched message doesn't guarantee parallel consumption.
> 
> With those three modes, I think it should cover all use case:

No it doesn't. Especially because you're focused on your use case.

I personally don't want to lock on consumption. I want to lock on dispatch.
I have process which dispatch lot of message often duplicate. Process is quick, but I don't need to plan processing 10 times the message.  When the first message is not processed the second one will give the same result, so is unneeded, then I don't want to add it to the queue.

I assume you didn't read the initial issue then https://github.com/symfony/symfony/issues/54126 it's talking about Deduplicating message IN QUEUE.

So I want a lock and discard message in queue, when YOU want a lock a discard on process ; 2 different use case.
People might also prefer to release the lock before consuming the message or after the consumption ; I use the 2 differents use case. So now we have 3 use case.

I copy this for lock and block to be explicit. I personnaly don't care bout this feature.

> Yet in the use case I had in the past, the non-blocking-and-requeue behavior is the one I needed the most

Still u never provide a PR for this feature and hijacking mine to add you custom need, when my PR came from a totally different issue. Uncool.

For the lock and retry, please open another PR. I won't list and implement every possible use case.

> but if we doesn't reach a consensus about it

Honestly after seeing this PR opened for 6 month and so much discussion about it ; the only consensus I see it the fact I can just close this PR cause it will not move forward

pounard replied:
@VincentLanglet

Ohhhhh OK my bad. Now I reread more thoroughly the original issue, I do better understand it. I do sincerely apologize for this, I was mistaken.

So I guess your original PR before I got here was the right one considering the issue topic. Now, can I make a suggestion? As your issue states, the initial goal is to **deduplicate** messages **in queue**.

Why wouldn't you simply rename the middleware to a less generic and conflicting name, such as **DeduplicateMiddleware**?

Also I would suggest renaming the attribute to simply **Deduplicate**?

I'm sorry for hijacking, but now I get that the name mislead me all along. Rename it, make it naive-simple just as you did in the first place, and explain better in code documentation what it really does and what is its purpose.

VincentLanglet replied:
Definetly a good idea ; let's go with one stamp/middleware by usecase. :)

VincentLanglet commented:
Hi @nicolas-grekas I'm unsure to know what is blocking in the Review process of this PR and don't know how to help moving it forward. What's missing ? Thanks

nicolas-grekas commented on src/Symfony/Component/Messenger/Middleware/LockMiddleware.php:
```diff
@@ -0,0 +1,54 @@
+
+    public function handle(Envelope $envelope, StackInterface $stack): Envelope
+    {
+        $stamp = $envelope->last(LockStamp::class);
+        if (null === $stamp) {
```
```suggestion
        if (!$stamp = $envelope->last(LockStamp::class)) {
```

nicolas-grekas commented on src/Symfony/Component/Messenger/Middleware/LockMiddleware.php:
```diff
@@ -0,0 +1,54 @@
+            return $stack->next()->handle($envelope, $stack);
+        }
+
+        if (null === $envelope->last(ReceivedStamp::class)) {
```
```suggestion
        if (!$envelope->last(ReceivedStamp::class)) {
```

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTestCase.php:
```diff
@@ -1060,25 +1061,54 @@ public function testMessengerWithMultipleBuses()
-            ['id' => 'handle_message', 'arguments' => [false]],
-        ], $container->getParameter('messenger.bus.commands.middleware'));
+
+        if (class_exists(LockMiddleware::class)) {
```
Don't we also need to check that the lock component is installed?

VincentLanglet replied:
It's installed, we're in the test directory and `symfony/lock` is in require-dev.

nicolas-grekas replied:
I mean on a user app. The current code will enable the middleware even if the lock component is not installed.

VincentLanglet replied:
We're on the class
```
src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTestCase.php
```
which is in the test folder so this is not installed in the user app.

On the other side, there is 
```
src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
```
which is in the user app, and for this one I have already checked 
```
if (class_exists(LockMiddleware::class) && class_exists(LockFactory::class)) {
```
cf
https://github.com/symfony/symfony/pull/54141/files#diff-e4f52d5033296421d02e360c8355aa1931bdfb0c12579f4815ea02b724322c0cR2179

nicolas-grekas replied:
Aaah sorry I missed the "Test" part in the classname...

nicolas-grekas commented on src/Symfony/Component/Messenger/Middleware/LockMiddleware.php:
```diff
@@ -0,0 +1,54 @@
+        try {
+            $envelope = $stack->next()->handle($envelope, $stack);
+        } finally {
+            if (
+                null !== $envelope->last(ReceivedStamp::class)
+                && !$stamp->shouldBeReleasedBeforeHandlerCall()
+            ) {
```
```suggestion
            if ($envelope->last(ReceivedStamp::class) && !$stamp->shouldBeReleasedBeforeHandlerCall()) {
```

nicolas-grekas commented:
Oh, and a line is missing in the changelog

pounard commented:
@VincentLanglet I want to apologize for having misunderstood the original issue. Please ignore all my comments.

May I suggest renaming the whole to something much more specific and explicit about what it does and why, such as @jderusse suggested in one the first comments?

Some suggestions about how I would name those classes:
- `LockMiddleware` -> `DeduplicateMiddleware`
- `LockStamp` -> `DeduplicateStamp`

Because the "Lock" term may mislead to "contention" or "conflict" problem resolution, such as deadlocks or transaction failures in database for example, but in my opinion doesn't ring any bell about deduplication.

pounard commented on src/Symfony/Component/Messenger/Middleware/DeduplicateMiddleware.php:
```diff
@@ -0,0 +1,51 @@
+            if ($envelope->last(ReceivedStamp::class) && !$stamp->onlyDeduplicateInQueue()) {
+                $this->lockFactory->createLockFromKey($stamp->getKey())->release();
+            }
+        }
```
So if I'm reading right, you have here two use cases:
- Only in queue means that the acquire then release happens while dispatching, am I right?
- The other one means that it acquires when dispatching, and release while receiving?

Will the second one be reliable in all cases? It seems that it's highly reliant on the fact that the lock is distributed and cannot fail. But it seems OK, it will prevent any messages from being sent while another is consumed as expected, I just want to be sure I understood it well.

VincentLanglet replied:
I'm not sure what you mean by dispatching.

To me there is three step
- I put the message in queue
- I receive the message from the queue to treat the message
- I finished the treatment

Only in queue mean I don't want to put two times the same message in queue because they will have the same effect. But since the treatment of the message take time and has multiple steps, even if a first message is in treatment, the following might have a different result so I have to add it to the queue.

The other one, means as soon as I have a message in queue or in treatment, I don't want to add another one.

welcoMattic reviewed (approved):
It look good to me. If possible to rebase the branch and squash commits @VincentLanglet, it will make the merge easier. Thanks for your contribution!

VincentLanglet committed:
Deduplicate Middleware

VincentLanglet commented:
> It look good to me. If possible to rebase the branch and squash commits @VincentLanglet, it will make the merge easier. Thanks for your contribution!

Rebased and squashed

ro0NL commented on src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTestCase.php:
```diff
@@ -1061,25 +1062,54 @@ public function testMessengerWithMultipleBuses()
-            ['id' => 'handle_message', 'arguments' => [false]],
-        ], $container->getParameter('messenger.bus.commands.middleware'));
+
+        if (class_exists(DeduplicateMiddleware::class)) {
+            $this->assertEquals([
+                ['id' => 'add_bus_name_stamp_middleware', 'arguments' => ['messenger.bus.commands']],
+                ['id' => 'reject_redelivered_message_middleware'],
+                ['id' => 'dispatch_after_current_bus'],
+                ['id' => 'failed_message_processing_middleware'],
+                ['id' => 'deduplicate_middleware'],
+                ['id' => 'send_message', 'arguments' => [true]],
+                ['id' => 'handle_message', 'arguments' => [false]],
+            ], $container->getParameter('messenger.bus.commands.middleware'));
+        } else {
```
update the test deps next time to avoid these exra conditions :)

ro0NL replied:
it's all maintenance

ro0NL replied:
cc @fabpot

xabbuh replied:
There’s not really a need to forbid using FrameworkBundle with Messenger < 7.3. But that would be the consequence if we cannot test it.

ro0NL replied:
i dont like the if/else condition based on class presence generally

ro0NL replied:
7.2 is tested without class X
7.3 is tested with class X

it's that simple to me.

xabbuh replied:
That would complicate things. Either we would require people to update FrameworkBundle and components at the same time which would make updating harder. Or we would not be able to test the bundle with the lowest deps which would not allow us to detect compatibility issues.

ro0NL replied:
> we would require people to update FrameworkBundle and components at the same time which would make updating harder

generally yes, it that prevents if/else flows in your codebase. We should prefer branches then.

updating many sf packages from 7.2 to 7.x is not an issue.
